### ISSUE TITLE
Support parallel test execution

### ISSUE THREAD

**Issue by kyriee** [FakeItEasy/FakeItEasy#60]
Parallel test execution is not currently supported.
If it turned on, ExpectationException (Call to non configured method) errors appear :(

**Comment by adamralph** [comment]
Now that xunit 2.0 has gone stable, and runs tests in parallel by default, this is a more pressing issue.

**Comment by blairconrad** [comment]
I didn't realize that XUnit was switching to running tests in parallel by default. An interesting choice. Given most of the tests I've seen in the wild, this would've been disastrous. Then again, the ones I usually see are written in NUnit, and I understand the patterns are different.

I worry that we'll be able to make this change in a timely fashion. In particular (and without diving into the code) I think the scopes support will be problematic.

**Comment by adamralph** [comment]
Actually I should qualify that: the default behaviour is to run all tests *within a class* in parallel. There is also an option to run all tests in an assembly in parallel.

**Comment by adamralph** [comment]
A sledge hammer approach to scopes (and perhaps the only feasible one) is to block any threads trying to create a new scope until the current on is disposed. This would effectively serialise tests which would otherwise be parallel.

**Comment by blairconrad** [comment]
At least a new non-nested scope. Which I think we support. Although I shouldn't comment without taking more time to look at the code than I'm going to this morning.
I don't hate that approach, if it's all we can do. Or a stop-gap.

**Comment by aarondandy** [comment]
Here is a related issue: https://github.com/xunit/xunit/issues/357

A large portion of my tests are and use async methods so this is an issue for me as well. For now I am marking my test assemblies with `[assembly: CollectionBehavior(DisableTestParallelization = true)]` (note that the spelling of the property may differ between versions of xunit).

**Comment by blairconrad** [comment]
*Very preliminary* testing supports the idea that the `That.Matches` is one failure point in the thread-unsafety. I have failing test cases based off https://gist.github.com/StefanBilliet/982ef274c4303235ecd7. Converting the tests to non-Tasks exhibits the same behaviour, but when I drop the `That.Matches`, I get no test failures.

I suspect that the static `ArgumentConstraintTrap.trappedConstraints` is the culprit. Will investigate more later in the weekend (family events interfere).

**Comment by GraemeBradbury** [comment]
To add to this. We've started seeing an issue where calls to faked methods are returning Dummies even though setups have been done.
eg: `A.CallTo(...).Return(...)` are being ignored.

Current instances seen have always been setups in xBehave `[Background]` steps so not yet sure if it's a FakeItEasy issue or a xBehave one.

**Comment by adamralph** [comment]
@GraemeBradbury I'd hazard a guess that it's almost certainly a FakeItEasy problem. xBehave 2.0 is well isolated in terms of thread safety.

**Comment by blairconrad** [comment]
@GraemeBradbury this is with xBehave 2.0, yes?
Which version of FakeItEasy? Have you a small sample you can share?

**Comment by GraemeBradbury** [comment]
Nuget Package versions:
xBehave 2.0.0
xUnit 2.0.0
FakeItEasy 1.25.3

I'm trying to create a minimum example test. The tests that exhibit the problem are all integration tests (faking out the file system). Will update as soon as I have something.

**Comment by blairconrad** [comment]
Ah. The FakeItEasy 2.0 betas definitely have some threadsafety improvements.
If your problems go away when all tests are run single-threaded under 1.25.3, I'd be inclined to move on to how they behave under the latest 2.0.
Speaking for myself, I've no interest in making 1.x.y threadsafer…

**Comment by adamralph** [comment]
:+1: for for giving it a try with `[assembly:Xunit.CollectionBehavior(MaxParallelThreads = 1)]`

**Comment by apacha** [comment]
Is there a roadmap when Version 2 will be released? We are looking forward to the support of parallel test-execution because our Jenkins is constantly turning red because of FakeItEasy, although we have tried to configure thread-safety during creation of the fake object, we now get race-conditions during `A.CallTo(...).MustHaveHappened()` - assertions because of parallel test-execution.

**Comment by blairconrad** [comment]
@apacha, we have no dates set. Since we work on FakeItEasy in our spare time, and have outside interests, it's nearly impossible to predict. The other answer is "as soon as we can", as we'd enjoy publishing the changes that we have and then resuming our regular cycle of frequent small releases.

I am surprised your tests are running in parallel, and have commented on your [StackOverflow question](http://stackoverflow.com/questions/33120648/run-nunit-tests-sequentially-on-a-jenkins-server). Perhaps we should meet up over there to discuss how your tests are running in parallel to begin with.

In the meantime, there are no "threadsafety" issues planned for 2.0, so 2.0.0-beta010 has everything that 2.0 is likely to get. You could try that release.
Or, if you can help identify a particular problem not addressed by 2.0.0-beta010, raise it and let's see what we can do.

**Comment by apacha** [comment]
Thanks for the fast response. We will try 2.0.0-beta010 to see if this resolves our issues.

**Comment by blairconrad** [comment]
:+1: @apacha. I'd be keen to hear how it works out for you.

**Comment by apacha** [comment]
I've tried 2.0.0-beta010 today and ran our test-suite (approx. 800 tests) about ten times with the newer version of FakeItEasy. My impression is that we get fewer race-conditions during creation and assertion, but they still persist ("Enumeration was changed" exceptions and A.CallTo()-assertion that were lost). We will investigate further how the tests are running on the Jenkins.

**Comment by blairconrad** [comment]
Very peculiar. I'd be happy to help take a look at whatever portions of the tests and code you can share, once you've a little more of a handle on what's going wrong.

**Comment by adamralph** [comment]
Note, via @dennisdoomen,  https://twitter.com/ddoomen/status/685108134269026304?s=09

> Just make sure you don't use Fake.CreateScope. That'll break it.

**Comment by blairconrad** [comment]
Yup, I also refer readers to #484.

**Comment by adamralph** [comment]
This has been taken care of in several other issues as part of 2.0, with the only known outstanding issue being https://github.com/FakeItEasy/FakeItEasy/issues/484.

If any other problems with parallel execution arise, we can deal with them as bugs at that time.

**Comment by erlis** [comment]
Team, 

I see a few thread safety issues but all of them have been marked as closed. 

I'm using version 2.3.3 and NUnit 3.4.1 running tests in parallel with the  `Parallelizable(ParallelScope.Fixtures)` setting and I'm having FIE failing an expectection with message: 

` Assertion failed for the following call:
    Common.Logging.IAppLogger.Error("DeleteEmployeesByX DeleteEmployee failed. EeId: {0}", [<Ignored>])
  Expected to find it exactly 2 times but found it #1 times among the calls: ....`

Is this because the **Ignored** parameter? 

The test passes if parallel execution is disabled.

Thanks, 
Erlis

**Comment by blairconrad** [comment]
Hi, @erlis. Thanks for bringing this up.

As you say, all of our known thread safety issues have been closed. It's entirely possible that we've missed something.
I doubt the `Ignored` is the culprit here (but anything's possible).

It's hard to do much debugging without more information. Do you have a minimal test case that reproduces the issue and that you could share? Alternatively, I can try to whip something up, but I won't be able to look at it for at least 2 days, and there's no guarantee that I'll be trying the right thing.

One thing I will ask, though. You're not sharing Fake objects between tests that are running in parallel, are you?

**Comment by erlis** [comment]
hi @blairconrad, 

I found that the error I was having was not because FakeItEasy but an actual bug on my code, I'm testing the parallel execution and so far everything is working as expected, If I find a problem.. should I open a new bug or add to this one? 

Regards, 
Erlis

**Comment by adamralph** [comment]
@erlis yes, if you find any problems, please open a new issue. Many thanks.

**Comment by blairconrad** [comment]
@erlis, sorry I never got back to you. Work and non-work things interfered (and somewhere in the back of my head probably a feeling that I'd never be able to reproduce your problem with the info I had). I'm glad things worked out for you, and that FakeItEasy, however indirectly, helped you find the bug in your code.
